File: frontend/olympic_it_project/lib/core/network/api_response.dart
class ApiResponse<T> {
    final int code;
    final String message;
    final T? data;

    ApiResponse({
        required this.code,
        required this.message,
        this.data
    });

    factory ApiResponse.fromJson(Map<String, dynamic> json, T Function(dynamic) fromJsonT) {
        return ApiResponse(
            code: json['code'] as int,
            message: json['message'] as String,
            data: json['data'] != null ? fromJsonT(json['data']) : null
        );
    }
}

File: admin-web/js/core/api_response.js
export function unwrap(json, fromJson) {
    if (!json) {
        throw new Error("Không có dữ liệu trả về từ server");
    }

    const { code, message, data } = json;
    if (code !== 200) {
        throw new Error(message ?? "Có lỗi xảy ra");
    }

    if (data === undefined || data === null) {
        return data;
    }

    return fromJson ? fromJson(data) : data;
}